### PR TITLE
Improve LSDK by using submodules and better makefile

### PR DESCRIPTION
**Pull request by jordanrh1**
This is a complete rewrite of LSDK to use submodules (for deterministic commit fetching) and a rewritten Makefile that's simpler, faster, and easier to understand.

### CODE REVIEW

**Comment by jiria** [comment]
- Added Hernan to make sure this has no implications to OE repo.
- Consider grouping external dependencies to cleanup the top level directory (say under \external or \deps.
- Tried to checkout on Windows under d:\Repos\jordan_lsdk and it seems that some subpaths are too long:

Submodule path 'cst': checked out '3285e1d23a6abe3db6b3673e79acdeed4e19ea95'
error: unable to create file drivers/gpu/drm/nouveau/nvkm/subdev/i2c/aux.c: No such file or directory
error: unable to create file drivers/gpu/drm/nouveau/nvkm/subdev/i2c/aux.h: No such file or directory
error: unable to create file include/soc/arc/aux.h: No such file or directory
Submodule path 'ms-tpm-20-ref': checked out 'd979cae6db01ea2e57dbce92d700ce1d0edf78fa'

Submodule path 'optee_os/external/RIoT/External/tinycbor': checked out 'aa3731592c784ce91e7fcafa66665c147b18be25'
error: unable to create file package/testsuite/global_platform/api_1.0/GP_XSL_TEE_Initial_Configuration-Test_Suite_v1_0_0-2014-12-03-STM/TTAs/TTA_ClientAPI/TTA_answerErrorTo_OpenSession/code_patches/v1_1_0_4-2014_11_07/TTA_answerErrorTo_OpenSession_protocol.h.patch: Filename too long
error: unable to create file package/testsuite/global_platform/api_1.0/GP_XSL_TEE_Initial_Configuration-Test_Suite_v1_0_0-2014-12-03-STM/TTAs/TTA_ClientAPI/TTA_answerSuccessTo_OpenSession_Invoke/code_patches/v1_1_0_4-2014_11_07/TTA_answerSuccessTo_OpenSession_Invoke_protocol.h.patch: Filename too long
error: unable to create file package/testsuite/global_platform/api_1.0/GP_XSL_TEE_Initial_Configuration-Test_Suite_v1_0_0-2014-12-03-STM/TTAs/TTA_ClientAPI/TTA_check_OpenSession_with_4_parameters/code_patches/v1_1_0_4-2014_11_07/TTA_check_OpenSession_with_4_parameters_protocol.h.patch: Filename too long
Submodule path 'ppa-generic': checked out '90e13c9e148972f75f5f2e7f7a904dabf1586049'

**Comment by jordanrh1** [comment]
What is OE using LSDK for?

**Comment by jiria** [comment]
@jordanrh1 , not saying it is; but Hernan has been working on cleaning up the Linux ARM build system so want to make sure there is no inferred dependency.

**Comment by jiria** [comment]
Attempting to build on Linux, the firmware fails to build due to missing dependency: bison. Installed that, the next missing dependency is flex. I remember the old instructions had some initialization. Is that still needed or should we no longer depend on flex?

**Comment by jordanrh1** [comment]
It's possible I'm missing dependencies in the instructions because my computer has so many packages installed. I'll add flex and bison to the 'Install Prerequisites' section.

**Comment by jiria** [comment]
Ok, that helped. Few more missing dependencies: libssl-dev and python. I was still missing pycrypto. For that, I had to install python-pip as well. That brought pycrypto on its own, so I did not have to use pip to install it. Not sure if python-pip is sufficient and python itself would be a transitive dependency of that (also did not check py 2 vs 3). At this point, make firmware is still failing with lsch2.h no such file or directory. Looking further...

**Comment by jiria** [comment]
It would seem that lsch2.h is part of the checked out tree, yet make firmware is not finding it for me. Rerunning make firmware, the error went away and I was able to build the firmware. Onto the next stage...

**Comment by jordanrh1** [comment]
Do you know which subtarget it was trying to build when it couldn't find `lsch2.h` ?

**Comment by jiria** [comment]
image

**Comment by jordanrh1** [comment]
Very weird. It looks like lsch2.h gets copied to the build/src directory. I am not able to reproduce this error.

**Comment by jiria** [comment]
I have done make clean and make firmware again and it does not repro. Possibly something to do with missing dependencies that were added later on.

**Comment by jiria** [comment]
This happened during make os, it seems that "install" is missing in this command:

image

**Comment by jiria** [comment]
And manually installing them onto the host does not help either:

image

The host is 18.04 Ubuntu.

**Comment by jordanrh1** [comment]
It looks like the qemu packages are only available on Ubuntu 18.10 and later.

**Comment by jordanrh1** [comment]
It looks like [these](https://www.reddit.com/r/Ubuntu/comments/948cc1/ok_i_give_up_how_do_i_install_qemu/) instructions will work to install qemu on your 18.04 system, but it is probably easier to do a dist-upgrade.

```
sudo apt update
sudo apt dist-upgrade
```

**Comment by jiria** [comment]
Ok, will retry on 18.10. Please make sure the instructions explicitly say that only 18.10 has been tested and is recommended. We should also say that we know that 18.04 will not work.

**Comment by jordanrh1** [comment]
Wait a second, my machine is running Ubuntu 18.04.1 LTS. I need to investigate why it's working on mine.

**Comment by jordanrh1** [comment]
`qemu-user-static` is available on 18.04 (bionic). https://packages.ubuntu.com/bionic/qemu-user-static
Can you try running
```bash
sudo apt update
sudo apt install qemu-user-static
```

**Comment by jiria** [comment]
Will need to start from scratch, the default template VM I have been using is out of space. Will try 18.04 and 18.10 in parallel based on the most recent instructions on this branch. Please let me know if you have any other feedback.

File: Readme.md [hunk 1]
```diff
@@ -1,12 +1,7 @@
 Getting started on LS1012 Grapeboard
 ============
 
-This document will walk you through building all components from source and developing OP-TEE TA's for [Scalys Grapeboard](https://www.grapeboard.com/). The components we will build are
-
- - RCW, PBL, and U-Boot
- - PPA (Primary Protected Application)
- - OP-TEE
- - Linux
+This document will walk you through building all components from source and developing OP-TEE TA's for [Scalys Grapeboard](https://www.grapeboard.com/). We will build firmware, linux image, and test applications for Scalys grapeboard.
```
Comment (jiria) [inline Readme.md]: Would suggest adding below that the physical box is needed for building of the code as well.

File: Readme.md [hunk 2]
```diff
@@ -62,104 +57,84 @@ sudo apt install build-essential gcc-aarch64-linux-gnu g++-aarch64-linux-gnu u-b
...
+Change directories to the root of this repository, and run
 
 ```
-git clone --recurse-submodules https://github.com/ms-iot/u-boot.git -b ms-iot-grapeboard
-cd u-boot
-export ARCH=aarch64
-export CROSS_COMPILE=aarch64-linux-gnu-
-make grapeboard_pcie_qspi_spl_defconfig
-make
+git submodule init
```
Comment (jiria) [inline Readme.md]: Imho there should be git clone command for this current repo as well, which can clone submodules recursively.

File: Readme.md [hunk 3]
```diff
@@ -62,104 +57,84 @@ sudo apt install build-essential gcc-aarch64-linux-gnu g++-aarch64-linux-gnu u-b
 
 ```
 
-# Building RCW, PBL, and U-Boot
+# Building Firmware
```
Comment (jiria) [inline Readme.md]: For the prereqs above, git is missing as well.

**Comment by jiria** [comment]
I do not think the qemu is the problem. The problem is that the command to install it is:

sudo apt-get –asume-yes \
                binfmt-support qemu-system-common qemu-user-static

Notice there is no “install” in the command above. After adding install and manually running the command, it works ok.

________________________________

**Comment by jiria** [comment]
I have added few comments for the instruction page inline.

To make firmware on both 18.04 and 18.10, it is sufficient to add the following packages to the prereq section:

git flex bison python-pip libssl-dev

To make os on 18.04, there is a command trying to install binfmt and qemu. That command is missing "install". That needs to be added. The make os however fails later with: 

image

I did not proceed further.

To make os on 18.10, it fails very early on with:

image

I did not proceed further.

**Comment by jordanrh1** [comment]
What's your compiler version?
```
aarch64-linux-gnu-gcc --version
aarch64-linux-gnu-gcc (Ubuntu/Linaro 7.3.0-27ubuntu1~18.04) 7.3.0
```

**Comment by jiria** [comment]
Same one on 18.04, 8.2.0 on 18.10.

Sent from Mail<https://go.microsoft.com/fwlink/?LinkId=550986> for Windows 10

________________________________

**Comment by jordanrh1** [comment]
@jiria I have fixed the issues on 18.04 - please try again. I have not fixed the 18.10 issues; don't bother with 18.10.

**Comment by jiria** [comment]
18.04 looks good now.

**Pull request merged**